In [7]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 8: I Heard You Like Registers ---</h2><p>You receive a <span title="There's that sorcery I told you about.">signal</span> directly from the CPU. Because of your recent assistance with <a href="5">jump instructions</a>, it would like you to compute the result of a series of unusual register instructions.</p>
<p>Each instruction consists of several parts: the register to modify, whether to increase or decrease that register's value, the amount by which to increase or decrease it, and a condition. If the condition fails, skip the instruction without modifying the register. The registers all start at <code>0</code>. The instructions look like this:</p>
<pre><code>b inc 5 if a &gt; 1
a inc 1 if b &lt; 5
c dec -10 if a &gt;= 1
c inc -20 if c == 10
</code></pre>
<p>These instructions would be processed as follows:</p>
<ul>
<li>Because <code>a</code> starts at <code>0</code>, it is not greater than <code>1</code>, and so <code>b</code> is not modified.</li>
<li><code>a</code> is increased by <code>1</code> (to <code>1</code>) because <code>b</code> is less than <code>5</code> (it is <code>0</code>).</li>
<li><code>c</code> is decreased by <code>-10</code> (to <code>10</code>) because <code>a</code> is now greater than or equal to <code>1</code> (it is <code>1</code>).</li>
<li><code>c</code> is increased by <code>-20</code> (to <code>-10</code>) because <code>c</code> is equal to <code>10</code>.</li>
</ul>
<p>After this process, the largest value in any register is <code>1</code>.</p>
<p>You might also encounter <code>&lt;=</code> (less than or equal to) or <code>!=</code> (not equal to). However, the CPU doesn't have the bandwidth to tell you what all the registers are named, and leaves that to you to determine.</p>
<p><em>What is the largest value in any register</em> after completing the instructions in your puzzle input?</p>
</article>


In [8]:
from operator import eq, ge, gt, le, lt, ne
from re import fullmatch
from typing import Generator

from more_itertools import first


example = """
b inc 5 if a > 1
a inc 1 if b < 5
c dec -10 if a >= 1
c inc -20 if c == 10
"""


def largest_value_register(instructions: str) -> int:
    operations = {">": gt, ">=": ge, "==": eq, "!=": ne, "<=": le, "<": lt}
    registers = defaultdict(int)

    maximum = -inf
    for l in instructions.strip().splitlines():
        r0, op, v0, _, r1, co, v1 = parse(l)

        if operations[co](registers[r1], v1):
            registers[r0] += v0 if op == "inc" else -v0
            maximum = max(maximum, registers[r0])

    return max(registers.values()), maximum


def parse(l: str) -> Generator[tuple[int, str, str, str, str, str, int]]:
    return (int(ch) if fullmatch(r"-?[0-9]+", ch) else ch for ch in l.split(" "))


assert first(largest_value_register(example)) == 1

In [9]:
with open("../input/day8.txt") as f:
    puzzle = f.read()

print(f"Part I: {first(largest_value_register(puzzle))}")

Part I: 4567


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>4567</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>To be safe, the CPU also needs to know <em>the highest value held in any register during this process</em> so that it can decide how much memory to allocate to these operations. For example, in the above instructions, the highest value ever held was <code>10</code> (in register <code>c</code> after the third instruction was evaluated).</p>
</article>

</main>


In [10]:
from more_itertools import last


assert last(largest_value_register(example)) == 10

In [11]:
with open("../input/day8.txt") as f:
    puzzle = f.read()

print(f"Part II: {last(largest_value_register(puzzle))}")

Part II: 5636


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>5636</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>


<link href="style.css" rel="stylesheet"></link>
